api/auth.go

```diff
@@ -61,16 +56,65 @@ func (a *API) withToken(ctx context.Context, w http.ResponseWriter, r *http.Requ
 		}
 	}
 
-	log = log.WithFields(logrus.Fields{
+	log.WithFields(logrus.Fields{
 		"claims_id":    claims.ID,
 		"claims_email": claims.Email,
 		"roles":        roles,
 		"is_admin":     isAdmin,
-	})
+	}).Info("successfully parsed claims")
 
-	log.Info("successfully parsed claims")
 	ctx = gcontext.WithAdminFlag(ctx, isAdmin)
-	ctx = gcontext.WithLogger(ctx, log)
+	ctx = gcontext.WithToken(ctx, token)
+	return ctx, nil
+}
 
-	return gcontext.WithToken(ctx, token)
+func authRequired(w http.ResponseWriter, r *http.Request) (context.Context, error) {
+	ctx := r.Context()
+	claims := gcontext.GetClaims(ctx)
+	if claims == nil {
+		return nil, unauthorizedError("No claims provided")
+	}
+
+	return ctx, nil
+}
+
+func adminRequired(w http.ResponseWriter, r *http.Request) (context.Context, error) {
+	ctx := r.Context()
+	claims := gcontext.GetClaims(ctx)
+	isAdmin := gcontext.IsAdmin(ctx)
+
+	if claims == nil || !isAdmin {
+		return nil, unauthorizedError("Admin permissions required")
+	}
+
+	logEntrySetField(r, "admin_id", claims.ID)
+	return ctx, nil
+}
+
+func ensureUserAccess(w http.ResponseWriter, r *http.Request) (context.Context, error) {
+	ctx := r.Context()
+	userID := getUserID(ctx)
+
+	// ensure userID matches authenticated user OR is admin
+	claims := gcontext.GetClaims(ctx)
+	isAdmin := gcontext.IsAdmin(ctx)
+	if claims.ID != userID && !isAdmin {
+		return nil, unauthorizedError("Can't access a different user unless you're an admin")
+	}
+	if isAdmin {
+		logEntrySetField(r, "admin_id", claims.ID)
+	}
+
+	return ctx, nil
+}
+
+func hasOrderAccess(ctx context.Context, order *models.Order) bool {
+	claims := gcontext.GetClaims(ctx)
+
+	if order.UserID != "" {
+		if claims == nil || (order.UserID != claims.ID && !gcontext.IsAdmin(ctx)) {
```

Comment:
you should be able to short circuit the code and simplify that boolean with a check at the top. 

```
if gcontext.IsAdmin(ctx) {
  return true
}
if order.UserID == "" {
  return true
}

claims := gcontext.GetClaims(ctx)
return claims == nil || order.UserID != claims.ID
```